handwritten/hexagon_arch_c/functions.c

```diff
@@ -0,0 +1,687 @@
+			if (addr == pi->addr) {
+				return p;
+			}
+		}
+	}
+	return NULL;
+}
+
+/**
+ * \brief Frees an instruction.
+ * 
+ * \param i The instruction to be freed.
+ */
+RZ_API void hex_insn_free(HexInsn *i) {
+	if (i) {
+		free(i);
+	}
+}
+
+/**
+ * \brief Frees an constant extender.
+ * 
+ * \param ce The constant extender to be freed.
+ */
+RZ_API void hex_const_ext_free(HexConstExt *ce) {
+	if (ce) {
+		free(ce);
+	}
+}
+
+/**
+ * \brief Get the index of a packet in the state
+ *
+ * \param state The state to operade on.
+ * \param p The packet whichs index should be determined.
+ * \return ut8 The index of the packet in the given state. UT8_MAX if it is not in the state.
+ */
+static ut8 get_state_pkt_index(HexState *state, const HexPkt *p) {
+	HexPkt *sp;
+	for (ut8 i = 0; i < HEXAGON_STATE_PKTS; ++i) {
+		sp = &state->pkts[i];
+		if (sp->pkt_addr == p->pkt_addr) {
+			return i;
+		}
+	}
+	return UT8_MAX;
+}
+
+/**
+ * \brief Initializes each packet of the state once.
+ * 
+ * \return The initialized state of the plugins.
+ */
+RZ_API HexState *hexagon_get_state() {
+	static HexState *state = NULL;
+	if (!state) {
+		state = calloc(1, sizeof(HexState));
+		if (!state) {
+			RZ_LOG_FATAL("Could not allocate memory for HexState!");
+		}
+		for (int i = 0; i < HEXAGON_STATE_PKTS; ++i) {
+			state->pkts[i].insn = rz_list_newf((RzListFree)hex_insn_free);
+			if (!state->pkts[i].insn) {
+				RZ_LOG_FATAL("Could not initilize instruction list!");
+			}
+			hex_clear_pkt(&(state->pkts[i]));
+		}
+		state->const_ext_l = rz_list_newf((RzListFree)hex_const_ext_free);
+	}
+	return state;
+}
+
+/**
+ * \brief Checks if the packet has 4 instructions set.
+ * 
+ * \param p The packet to check.
+ * \return true The packet stores already 4 instructions.
+ * \return false The packet stores less than 4 instructions.
+ */
+static inline bool is_pkt_full(const HexPkt *p) {
+	return rz_list_length(p->insn) >= 4;
+}
+
+/**
+ * \brief Sets the packet related information in an instruction.
+ * 
+ * \param hi The instruction.
+ * \param p The packet the instruction belongs to.
+ * \param k The index of the instruction wihin the packet.
+ */
+static void hex_set_pkt_info(const RzAsm *rz_asm, RZ_INOUT HexInsn *hi, const HexPkt *p, const ut8 k, const bool update_mnemonic) {
+	rz_return_if_fail(hi && p);
+	bool is_first = (k == 0);
+	HexPktInfo *hi_pi = &hi->pkt_info;
+
+	strncpy(hi_pi->syntax_postfix, "", 16);
+	// Parse instr. position in pkt
+	if (is_first && is_last_instr(hi->parse_bits)) { // Single instruction packet.
+		hi_pi->first_insn = true;
+		hi_pi->last_insn = true;
+		// TODO No indent in visual mode for "[" without spaces.
+		if (p->is_valid) {
+			strncpy(hi_pi->syntax_prefix, "[   ", 8);
+		} else {
+			strncpy(hi_pi->syntax_prefix, "?   ", 8);
+		}
+	} else if (is_first) {
+		hi_pi->first_insn = true;
+		hi_pi->last_insn = false;
+		if (p->is_valid) {
+			strncpy(hi_pi->syntax_prefix, rz_asm->utf8 ? "┌   " : "/   ", 8);
```

Comment:
Maybe make these special characters defines at the top of the file or in some headers? Like in Rizin?